File: packages/cli/src/commands/status.ts
import { $ } from "bun";
import { log } from "../lib/output";

const CONTAINERS = [
  { name: "infra-postgres", service: "PostgreSQL" },
  { name: "infra-redis", service: "Redis" },
  { name: "infra-rabbitmq", service: "RabbitMQ" },
];

export async function statusCommand(): Promise<void> {
  log.title("Infrastructure Status\n");

  for (const { name, service } of CONTAINERS) {
    try {
      const result =
        await $`docker inspect --format={{.State.Status}} ${name}`
          .nothrow()
          .quiet();

      const status = result.stdout.toString().trim();

      if (result.exitCode !== 0) {
        console.log(`  ${service.padEnd(14)} ⏹  not created`);
      } else if (status === "running") {
        console.log(`  ${service.padEnd(14)} 🟢 running`);
      } else {
        console.log(`  ${service.padEnd(14)} 🔴 ${status}`);
      }
    } catch {
      console.log(`  ${service.padEnd(14)} ⏹  not found`);
    }
  }

  console.log("");
}

File: packages/app/internal/docker/status.go
package docker

import (
	"context"
	"os/exec"
	"strings"
	"time"
)

type Status struct {
	Service   string
	Container string
	State     string
	Exists    bool
	Running   bool
}

var containers = []Status{
	{Service: "PostgreSQL", Container: "infra-postgres"},
	{Service: "Redis", Container: "infra-redis"},
	{Service: "RabbitMQ", Container: "infra-rabbitmq"},
	{Service: "AIStor", Container: "infra-aistor"},
}

func ListStatus(ctx context.Context) []Status {
	statuses := make([]Status, 0, len(containers))
	for _, item := range containers {
		statuses = append(statuses, inspect(ctx, item))
	}
	return statuses
}

func inspect(ctx context.Context, item Status) Status {
	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "docker", "inspect", "--format={{.State.Status}}", item.Container)
	output, err := cmd.CombinedOutput()
	if err != nil {
		item.State = "not created"
		item.Exists = false
		item.Running = false
		return item
	}

	item.State = strings.TrimSpace(string(output))
	item.Exists = true
	item.Running = item.State == "running"
	return item
}
